traffic_router/ultimate-test-harness/main_test.go

```diff
@@ -0,0 +1,160 @@
+var DNSRequestsPerSecondThreshold *int
+var BenchTime *int
+var ThreadCount *int
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+	IPv4Only = flag.Bool("4", false, "test IPv4 addresses only")
+	IPv6Only = flag.Bool("6", false, "test IPv4 addresses only")
+	CDNName = flag.String("cdn", "", "the name of a CDN to search for Delivery Services")
+	DeliveryServiceName = flag.String("ds", "", "the name (XMLID) of a Delivery Service to use for tests")
+	TrafficRouterName = flag.String("hostname", "", "the hostname of a Traffic Router to use")
+	UseCoverageZone = flag.Bool("coverage_zone", false, "whether to use an IP address from the Traffic Router's Coverage Zone File")
+	CoverageZoneLocation = flag.String("coverage_zone_location", "", "the Coverage Zone location to use (implies coverage_zone=true)")
+	HTTPRequestsPerSecondThreshold = flag.Int("http_requests_threshold", 8000, "the minimum number of HTTP requests per second a Traffic Router must successfully respond to")
+	DNSRequestsPerSecondThreshold = flag.Int("dns_requests_threshold", 25000, "the minimum number of DNS requests per second a Traffic Router must successfully respond to")
+	BenchTime = flag.Int("benchmark_time", 10, "the duration of each load test in seconds")
+	ThreadCount = flag.Int("thread_count", 12, "the number of threads to use for each test")
+
+	log.Init(os.Stderr, os.Stderr, os.Stderr, os.Stderr, os.Stderr)
+}
+
+func GetTrafficRouterDetails(t *testing.T) []TRDetails {
+	var err error
+	var trafficRouterDetails []TRDetails
+
+	ipAddressMaps := map[tc.CDNName]IPAddressMap{}
+
+	trafficRouters, err := getTrafficRouters(*TrafficRouterName, tc.CDNName(*CDNName))
+	if err != nil {
+		t.Fatalf("could not get Traffic Routers: %s", err.Error())
+	}
+
+	for _, trafficRouter := range trafficRouters {
+		var ipAddresses []string
+		for _, serverInterface := range trafficRouter.Interfaces {
+			if !serverInterface.Monitor {
+				log.Warnf("skipping server interface %s of Traffic Router %s because it is unmonitored", serverInterface.Name, *trafficRouter.HostName)
+				continue
+			}
+			ipv4, ipv6 := serverInterface.GetDefaultAddress()
+			if ipv4 != "" && !*IPv6Only {
+				ipAddresses = append(ipAddresses, ipv4)
+			}
+			if ipv6 != "" && !*IPv4Only {
+				ipAddresses = append(ipAddresses, "["+ipv6+"]")
+			}
+		}
+		if len(ipAddresses) < 1 {
+			log.Warnf("need at least 1 monitored service address on an interface of Traffic Router '%s' to use it for benchmarks, but %d such addresses were found", *trafficRouter.HostName, len(ipAddresses))
+			continue
+		}
+		dsTypeName := tc.DSTypeHTTP
+		httpDSes := getDSes(t, *trafficRouter.CDNID, dsTypeName, tc.DeliveryServiceName(*DeliveryServiceName))
+		if len(httpDSes) < 1 {
+			log.Warnf("at least 1 Delivery Service with type '%s' is required to run HTTP load tests on Traffic Router '%s', but %d were found", dsTypeName, *trafficRouter.HostName, len(httpDSes))
+		}
+		dsURL, err := url.Parse(httpDSes[0].ExampleURLs[0])
```

Comment:
This will panic if httpDSes is empty, and since that line is just a `warn`, it won't exit.

Reply:
Also, might be a good idea to add a `len` check to `ExampleURLs`

Reply:
> This will panic if httpDSes is empty, and since that line is just a `warn`, it won't exit.

`continue`ing if `httpDSes` is empty in 29bb35cd10

Reply:
> Also, might be a good idea to add a `len` check to `ExampleURLs`

Added a `len` check to `ExampleURLs` in 905db752c5